src/main/java/tech/picnic/jolo/Entity.java

```diff
@@ -69,37 +96,56 @@ private Entity(Table<R> table, Class<T> type, Field<Long> primaryKey, Field<?>[]
    * line. The corresponding class has one constructor without, and one with an argument {@code int
    * picked}. If it is mapped from a record that contains a (computed) {@code picked} column, then
    * the constructor with the extra argument is used to bring it into Java land.
+   *
+   * @apiNote This method returns a new object.
    */
   public Entity<T, R> withExtraFields(Field<?>... extraFields) {
-    this.fields = concat(stream(this.fields), stream(extraFields)).toArray(Field<?>[]::new);
-    return this;
-  }
-
-  /** Copies this entity, discarding any state. This method is used in a prototype pattern. */
-  public Entity<T, R> copy() {
-    return new Entity<>(table, type, primaryKey, fields);
+    Field<?>[] extendedFields =
+        concat(stream(fields), stream(extraFields).filter(Objects::nonNull))
+            .toArray(Field<?>[]::new);
+    return new Entity<>(table, type, primaryKey, extendedFields);
   }
 
   /** The table that is mapped by this entity. */
   public Table<R> getTable() {
     return table;
   }
 
-  /** The primary key that this Entity uses to distinguish records. */
+  /** The primary key that this entity uses to distinguish records. */
   public Field<Long> getPrimaryKey() {
     return primaryKey;
   }
 
+  /** The given record's ID, corresponding to this entity. */
+  Optional<Long> getId(Record record) {
+    check(record);
+    return Optional.ofNullable(record.get(primaryKey));
+  }
+
+  /** Loads an object of type {@code T} from the given record. */
+  T load(Record record) {
+    check(record);
+    /*
+     * The .into(resultFields) makes sure we don't let jOOQ magically use values from fields
+     * not included in `this.fields`. E.g., if `this.fields = [FOO.ID, FOO.X]` and the
+     * record contains FOO.ID=1 and BAR.X=1, then without this measure, BAR.X would be used
+     * instead of FOO.X.
+     */
+    T result = record.into(resultFields).into(type);
+    Objects.requireNonNull(result);
+    return result;
+  }
+
   /**
-   * Attempts to load an object of type {@code T} from the given record. If the {@link
-   * #getPrimaryKey() primary key} is not present in the record, no object is loaded. If the primary
-   * key is found in this record, and an object was already loaded for the same key, then the
-   * first-loaded object for this key is used.
+   * Check that the given record contains all expected fields and sets them.
+   *
+   * @implNote For performance reasons, this check is only performed for the first loaded record and
+   *     is a no-op otherwise.
    */
-  public void load(Record record) {
+  private void check(Record record) {
```

Comment:
I think this method no longer belongs here. It's keeping state about the data set being loaded inside the object.

Reply:
Why do we even need `resultFields`? Shouldn't it always be equivalent to `fields`? 
A mismatch IMO indicates that the loader definition doesn't match what the query returns and thus means some columns are `SELECT`ed but unused. Following best practices that is discouraged.